merkle-tree/concurrent/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+    pub fn root_index(&self) -> Result<usize, HasherError> {
+        let root = self.root();
+        for root_index in 0..(self.current_index as usize) + 1 {
+            if root == self.roots[root_index] {
+                return Ok(root_index);
+            }
+        }
+        Err(HasherError::RootNotFound)
+    }
+
+    fn inc_current_index(&mut self) {
+        self.current_index = (self.current_index + 1) % MAX_ROOTS as u64;
+    }
+
+    fn update_path_to_leaf(
+        &mut self,
+        leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]],
+    ) -> Result<(), HasherError> {
+        let mut leaf = *leaf;
+        let mut changelog_path = [[0u8; 32]; MAX_HEIGHT];
+
+        for (j, sibling) in proof.iter().enumerate() {
+            changelog_path[j] = leaf;
+            leaf = compute_parent_node::<H>(&leaf, sibling, leaf_index, j)?;
+        }
+
+        let changelog = ChangelogEntry::new(leaf, changelog_path, leaf_index);
+        self.inc_current_index();
+        self.changelog[self.current_index as usize] = changelog;
+        self.roots[self.current_index as usize] = leaf;
+
+        Ok(())
+    }
+
+    /// Replaces the `old_leaf` under the `leaf_index` with a `new_leaf`, using
+    /// the given `root` and `proof`.
+    pub fn update(
+        &mut self,
+        root_index: usize,
+        old_leaf: &[u8; 32],
+        new_leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]],
+    ) -> Result<(), HasherError> {
+        let proof: [[u8; 32]; MAX_HEIGHT] = if proof.len() < MAX_HEIGHT {
+            full_proof::<H, MAX_HEIGHT>(proof)
+        } else {
+            proof.try_into().map_err(|_| HasherError::ProofTooLarge)?
+        };
+
+        let mut updated_leaf = *old_leaf;
+        let mut updated_proof = proof.clone();
+        let mut k = root_index;
+        while k != self.current_index as usize {
+            let changelog_entry = self.changelog[k];
+            if leaf_index != changelog_entry.index as usize {
+                // This bit math is used to identify which node in the proof
+                // we need to swap for a corresponding node in a saved change log
+                let padding = 64 - MAX_HEIGHT;
+                let common_path_len = ((leaf_index ^ changelog_entry.index as usize) << padding)
+                    .leading_zeros() as usize;
+                let critbit_index = (MAX_HEIGHT - 1) - common_path_len;
+
+                updated_proof[critbit_index] = changelog_entry.path[critbit_index];
+            } else {
+                updated_leaf = changelog_entry.path[0];
+            }
+
+            k = (k + 1) % MAX_ROOTS;
+        }
+        if is_valid_proof::<H, MAX_HEIGHT>(
+            &self.roots[self.current_index as usize],
+            old_leaf,
+            leaf_index,
+            &proof,
+        )? && updated_leaf == *old_leaf
+        {
+            self.update_path_to_leaf(new_leaf, leaf_index, &updated_proof)?
```

Comment:
what' about calling the update_path_to_leaf method update_leaf.

Also I think this method would be easier to understand by encapsulating the code a bit differently:

```
pub fn update(
        &mut self,
        root_index: usize,
        old_leaf: &[u8; 32],
        new_leaf: &[u8; 32],
        leaf_index: usize,
        proof: &[[u8; 32]],
    ) -> Result<(), HasherError> {
        let proof: [[u8; 32]; MAX_HEIGHT] = if proof.len() < MAX_HEIGHT {
            full_proof::<H, MAX_HEIGHT>(proof)
        } else {
            proof.try_into().map_err(|_| HasherError::ProofTooLarge)?
        };
        
        let updated_proof = update_proof()
        
        is_valid_proof()?
        
        if updated_leaf != *old_leaf 
          retunr Err();
          
         self.update_leaf()
}

Reply:
I know this is super opinionated, just trying to make it as easy to understand as possible.

Reply:
What do you think?

Reply:
I like the idea. The only reason why I named the nested method `update_path_to_leaf` is that it was called so in the whitepaper.

And I especially like the idea of moving the proof update to a separate function.